redux-form/lib/reduxForm.d.ts

```diff
@@ -0,0 +1,551 @@
+     */
+    reduxMountPoint?: string;
+
+    /**
+     * An optional function you may provide to have full control over when sync validation happens.
+     */
+    shouldValidate?(params: ValidateCallback<FormData>): boolean;
+
+    /**
+     * An optional function you may provide to have full control over when async
+     * validation happens.
+     */
+    shouldAsyncValidate?(params: AsyncValidateCallback<FormData>): boolean;
+
+    /**
+     * marks fields as touched when the blur action is fired. Defaults to true.
+     */
+    touchOnBlur?: boolean;
+
+    /**
+     * marks fields as touched when the change action is fired. Defaults to
+     * false.
+     */
+    touchOnChange?: boolean;
+
+    /**
+     * a synchronous validation function that takes the form values and props
+     * passed into your component. If validation passes, it should return {}.
+     * If validation fails, it should return the validation errors in the form
+     * { field1: <String>, field2: <String> }.
+     * Defaults to (values, props) => ({}).
+     */
+    validate?(values: FormData, props: FormProps<FormData, P, S> & P): FormErrors<FormData>;
+
+    /**
+     * A synchronous warning function that takes the form values and props passed into your component.
+     * Warnings work the same as validations, but do not mark a form as invalid. If the warning check passes,
+     * it should return {}. If the check fails, it should return the warnings in the form
+     * { field1: <String>, field2: <String> }. Defaults to (values, props) => ({}).
+     */
+    warn?(values: FormData, props: FormProps<FormData, P, S> & P): FormWarnings<FormData>;
+}
+
+/**
+ * If your onSubmit function returns a promise, the submitting property will be set to true
+ * until the promise has been resolved or rejected. If it is rejected with a redux-form
+ * SubmissionError containing errors in the form { field1: 'error', field2: 'error' } then
+ * the submission errors will be added to each field (to the error prop) just like async
+ * validation errors are. If there is an error that is not specific to any field, but applicable
+ * to the entire form, you may pass that as if it were the error for a field called _error,
+ * and it will be given as the error prop.
+ */
+export interface SubmitHandler<FormData extends DataShape, P, S> {
+    (values: FormData, dispatch: Dispatch<S>, props: FormProps<FormData, P, S> & P): void | FormErrors<FormData> | Promise<any>;
+}
+
+interface ValidateCallback<FormData extends DataShape> {
+    /**
+     * The values.
+     */
+    values: FormData;
+
+    /**
+     * The next props.
+     */
+    nextProps: Object;
+
+    /**
+     * The current props.
+     */
+    props: Object;
+
+    /**
+     * true if the form is being initially rendered.
+     */
+    initialRender: boolean;
+
+    /**
+     * The structure object being used internally for values. You may wish to use 'deepEqual' from the structure.
+     */
+    structure: Object;
```

Comment:
You might want to define structure for plain and immutable such as here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13434/files#diff-92643af80dfab655413729df4df6fd6bR114
It's useful to have it typed.